nidirect_landing_pages/nidirect_landing_pages.module

```diff
@@ -42,7 +43,129 @@ function nidirect_landing_pages_entity_update(EntityInterface $entity) {
+    $subtheme = $user_input['field_subtheme'];
+  }
+  else {
+    return;
+  }
+
+  // Get a list of terms and articles that match this subtheme.
+  $teasers = nidirect_landing_pages_article_teasers($subtheme, $entity->id());
+
+  // Populate field_listing with paragraphs.
+  $langcode = $form_state->get('langcode');
+  $paras = [];
+  foreach ($teasers as $title => $atts) {
+    $paragraph = Paragraph::create([
+      'type' => $atts['type'],
+      'langcode' => $langcode,
+      $atts['field'] => ['target_id' => $atts['target_id']],
+    ]);
+    $paragraph->save();
+    $paras[] = ['target_id' => $paragraph->id(), 'target_revision_id' => $paragraph->getRevisionId()];
+  }
+  $entity->field_listing = $paras;
+}
+
+/**
+ * Utility function to get teasers.
+ */
+function nidirect_landing_pages_article_teasers(string $tid, $nid_to_exclude) {
+  $results = [];
+  // Get a list of article teasers by term.
+  $results = nidirect_landing_pages_get_article_teasers_by_term($tid, $nid_to_exclude);
+  // Get a list of article teasers by topic.
+  $results += nidirect_landing_pages_get_article_teasers_by_topic($tid);
+  // Alpha sort.
+  ksort($results);
+  return $results;
+}
+
+/**
+ * Utility function to get teasers by term.
+ */
+function nidirect_landing_pages_get_article_teasers_by_term(int $tid, $nid_to_exclude) {
+  // Render the 'articles by term' view and process the results.
+  $results = [];
+  $articles_view = views_embed_view('articles_by_term', 'article_teasers_by_term_embed', $tid, $tid);
+  \Drupal::service('renderer')->renderRoot($articles_view);
+  foreach ($articles_view['view_build']['#view']->result as $row) {
+    // This will be a link to an article.
+    if ($nid_to_exclude != $row->nid) {
+      $results[strtolower($row->_entity->getTitle())] = [
+        'type' => 'article_teaser',
+        'field' => 'field_article',
+        'target_id' => $row->nid,
+      ];
+    }
+  }
+  return $results;
+}
+
+/**
+ * Utility function to get teasers by topic.
+ */
+function nidirect_landing_pages_get_article_teasers_by_topic(int $tid) {
+  $results = [];
+  $articles_view = views_embed_view('site_subtopics', 'subtopic_teasers_by_topic_embed', $tid, $tid);
+  \Drupal::service('renderer')->renderRoot($articles_view);
+  foreach ($articles_view['view_build']['#view']->result as $row) {
+    // This will be a link to a taxonomy term.
+    $results[strtolower($row->_entity->getName())] = [
+      'type' => 'term_teaser',
+      'field' => 'field_term',
+      'target_id' => $row->tid,
+    ];
+  }
+  return $results;
+}
+
+/**
+ * Implements hook_page_attachments().
+ */
+function nidirect_landing_pages_page_attachments(array &$attachments) {
+  // Add landing page js if we are creating or editing
+  // a landing page node.
+  $add_js = FALSE;
+  if (\Drupal::service('router.admin_context')->isAdminRoute()) {
+    if (\Drupal::request()->server->get('REQUEST_URI') == '/node/add/landing_page') {
```

Comment:
Would be more robust to use the route name, but not a blocking issue.

Reply:
yes, good idea, have re-jigged the code to use route name now, which does feel much better